src/garble.js

```diff
@@ -1,182 +1,160 @@
- * @param {Object} wireToLabels - Mapping from each wire index to two labels
- * @param {Object} outputWireToLabels - Mapping from each output wire index to one label
- * @returns {number[]} Bit vector corresponding to the determined output
- */
-function outputLabelsToBits(circuit, wireToLabels, outputWireToLabels) {
-  var output = [];
-  for (var i = 0; i < circuit.wire_out_count; i++) {
-    var labelsForFalseAndTrue =
-      wireToLabels
-        .get(circuit.wire_out_index[i])
-        .map(function (l) { return l.withoutLastElement(); }); // Drop last bits.
-
-    var outputLabel =
-      outputWireToLabels
-        .get(circuit.wire_out_index[i])
-        [0] // Only one label in assignment from evaluator.
-        .withoutLastElement(); // Drop last bit.
-
-    var bit = outputLabel.getOccurrenceIndexIn(labelsForFalseAndTrue);
-    output.push(bit);
+const degarbleOutput = function (agent, garbledAssignment, outputLabels) {
+  const bits = [];
+  for (let i = 0; i < agent.circuit.outputSize; i++) {
+    const bitLabel = outputLabels[i];
+    const options = garbledAssignment[agent.circuit.wiresCount - agent.circuit.outputSize + i];
+    if (options[0].equals(bitLabel)) {
+      bits.push(0);
+    } else if (options[1].equals(bitLabel)) {
+      bits.push(1);
+    } else {
+      agent.progress('error', null, null, 'output label unequal to either labels');
+      agent.log('Output label ' + i + ' unequal to either labels!');
+    }
   }
-  return output;
-}
-
-/**
- * Garble all the gates (stateless version).
- * @param {Object} circuit - Circuit in which to garble the gates
- * @param {Object} wireToLabels - Mapping from each wire index to two labels
- * @returns {Object} Ordered collection of garbled gates
- */
-function garbleGates(circuit, wireToLabels) {
-  var gatesGarbled = new gate.GatesGarbled();
-  for (var i = 0; i < circuit.gate_count; i++) {
-    gatesGarbled.set(i, garbleGate(i, circuit.gate[i], wireToLabels));
+
+  return bits;
+};
+
+const run = function (agent) {
+  agent.progress('garbling', 0, agent.circuit.gates.length);
+
+  const circuit = agent.circuit;
+  const garbledCircuit = new Circuit(circuit.wiresCount, circuit.garblerInputSize, circuit.evaluatorInputSize, circuit.outputSize, agent.labelSize);
+
+  // generate random offset
+  const R = new Label(sodium.randombytes_buf(agent.labelSize));
+  R.setPoint(1);
+
+  // generate labels for input wires
+  const garbledAssignment = generateInputLabels(R, circuit.evaluatorInputSize + circuit.garblerInputSize, agent.labelSize);
+
+  // garble gates
+  for (let i = 0; i < circuit.gates.length; i++) {
+    let gate = circuit.gates[i];
+
+    if (gate.operation === 'AND') {
+      gate = garbleAnd(agent, gate, R, garbledAssignment);
+    } else if (gate.operation === 'XOR') {
+      gate = garbleXor(agent, gate, R, garbledAssignment);
+    } else {
+      gate = garbleInv(agent, gate, R, garbledAssignment);
+    }
+
+    garbledCircuit.gates.push(gate);
   }
```

Comment:
This does actually stall Chrome for a second or two, but at the moment I see no real need to check `this.throttle` until we aim to support slower/older JavaScript engines in the future.

Reply:
I think running this in the background using a web worker is probably the way to go, this can be done at the client side, not in JIGG, for flexibility.

https://www.w3schools.com/html/html5_webworkers.asp

I will add this to the client.html file in demo later